onnxruntime/core/graph/graph.cc

```diff
@@ -5537,6 +5537,42 @@ Graph::Graph(const Model& owning_model,
       is_loaded_from_model_file_(true) {  // true as the Graph isn't manually constructed from scratch
 }
 
+size_t Graph::ComputeNodeMemoryUsage(NodeIndex node_idx) const {
+  /// XXX: In some cases some kernels can copy its attributes to a device
+  // those are edge cases which we currently do not account for.
+  const Node* node = GetNode(node_idx);
+  if (node != nullptr) {
+    SafeInt<size_t> result = 0;
+    for (const auto* input : node->InputDefs()) {
+      if (input->Exists()) {
+        // Let's see if this is an initializer
+        constexpr const bool check_outer_scope_true = true;
+        const ONNX_NAMESPACE::TensorProto* initializer =
+            GetConstantInitializer(input->Name(), check_outer_scope_true);
+        if (initializer != nullptr) {
+          size_t out;
+          if (utils::GetSizeInBytesFromTensorProto<0>(*initializer, &out).IsOK()) {
+            result += out;
+          }
+        } else {
+          const auto* proto = input->TypeAsProto();
+          if (proto != nullptr && utils::HasTensorType(*proto)) {
+            const auto& tensor_type = proto->tensor_type();
+            if (utils::HasElemType(tensor_type) && utils::HasShape(tensor_type)) {
```

Comment:
What happens when an input does not have shape? Shall we add a log here (and seems result is not accurate in such case).

Reply:
Here and above. 
This function was created to compute the resource usage based on the graph alone in the absence of pre-recorded node stats. This would work if there are no dynamic shapes.
Currently, the threshold is optional. We can make also pre-recorded stats optional for a fully automatic mode when the threshold is automatically figured out based on the available memory. 
Alternatively, we can remove this and require pre-recorded profiling information for the partitioning to work.